src/psij/executors/batch/batch_scheduler_executor.py

```diff
@@ -0,0 +1,539 @@
+                if status.message is None:
+                    # only read output from submit script if another error message is not
+                    # already present
+                    status.message = self._read_aux_file(job, '.out')
+            else:
+                self._delete_aux_file(job, '.out')
+
+        except Exception as ex:
+            logger.warning('Job %s: failed to read auxiliary files: %s' % (job.id, ex))
+
+    def _read_aux_file(self, job: Job, suffix: str) -> Optional[str]:
+        assert job.native_id
+        path = self.work_directory / (job.native_id + suffix)
+        if path.exists():
+            try:
+                with open(path) as f:
+                    return f.read()
+            finally:
+                assert isinstance(self.config, BatchSchedulerExecutorConfig)
+                if not self.config.keep_files:
+                    path.unlink()
+        else:
+            return None
+
+    def _delete_aux_file(self, job: Job, suffix: str) -> None:
+        assert job.native_id
+        path = self.work_directory / (job.native_id + suffix)
+        if path.exists():
+            path.unlink()
+
+    def list(self) -> List[str]:
+        """Returns a list of jobs known to the underlying implementation.
+
+        See :func:`~JobExecutor.list`.
+        The returned list is a list of `native_id` strings representing jobs known to the
+        underlying batch scheduler implementation, whether submitted through this executor or not.
+        Implementations are encouraged to restrict the results to jobs accessible by the current
+        user.
+        """
+        pass
+
+
+class _QueuePollThread(Thread):
+    def __init__(self, name: str, config: BatchSchedulerExecutorConfig,
+                 executor: BatchSchedulerExecutor):
+        super().__init__()
+        self.name = name
+        self.daemon = True
+        self.config = config
+        self.executor = executor
+        # native_id -> job
+        self._jobs = {}  # type: Dict[str, Job]
+        # counts consecutive errors while invoking qstat or equivalent
+        self._poll_error_count = 0
+        self._jobs_lock = RLock()
+
+    def run(self) -> None:
+        logger.debug('Executor %s: queue poll thread started', self.executor)
+        while True:
+            self._poll()
+            time.sleep(self.config.queue_polling_interval)
+
+    def _poll(self) -> None:
+        with self._jobs_lock:
+            if len(self._jobs) == 0:
+                return
+        logger.info('Polling for %s jobs', len(self._jobs))
+        try:
+            out = self.executor._run_command(self.executor.get_status_command(self._jobs.values()))
+            logger.debug('Output from qstat: %s', out)
+            self._poll_error_count = 0
+
+            status_map = self.executor.parse_status_output(out)
+
+            with self._jobs_lock:
+                jobs_copy = dict(self._jobs)
+
+            for native_id, job in jobs_copy.items():
```

Comment:
Both the mapping in `self._jobs` and the one in `status_map` will conflict if two jpsi job instances point to the same native job.  This could be avoided by mapping to the jpsi job IDs (creating the status command then needs one additional lookup step).

Reply:
Another very good point.

Reply:
Should be fixed now. A list of jobs is kept for each native_id and most operations are done on the native_id list rather than the jobs (which may have duplicate native_ids).